### DISCUSSION TITLE
A rundown of the pluggable parts.

### DISCUSSION THREAD

**Discussion by ValdemarGr** [http4s/http4s#4690] [Q&A] [answered]
I cannot seem to find any resource documenting the implications of each module.

Why are there two custom built http clients and servers (blaze and ember) and how do they differ?
What are the stability, performance and feature set guarantees for various backends (blaze, ember, jetty, okhttp, async-http, tomcat ...)?

Thank you.

**Comment by ScalaWilliam** [comment, marked as answer]
Great question!

http4s is a library rather than framework and all of these that you mention are adapters to different frameworks/backends. What you find often in existing systems is that they already have a standard container and HTTP client and you really don’t want to add another one to maintain and configure (there simply is no consistency out there because all of them offer different feature sets in different ways). I have deployed http4s apps with a single API to both servlet containers like Jetty where a Servlet environment is appropriate and likewise using raw Blaze apps.

As for their own differences, a holistic comparison would be difficult to make however a few opinionated pointers would definitely help users. I would keep it simple and use Blaze by default unless you have some explicit needs.

Will see if I can write something up later on. I had written some articles on Scala and this could be a good candidate.

**Reply by rossabaker** [reply]
I've been meaning to throw together a feature matrix for the client and server backends for a long time, and something keeps coming up.  At a very high level:

## Servers

* blaze-server: oldest, most battle tested, supports web sockets, no new features
* ember-server: only one without `ConcurrentEffect` in CE2, graceful shutdown, easiest to maintain, HTTP/2 under active development
* jetty, tomcat: mature underlying server, sparsely used http4s integration.  Great way to pair FP Scala with Java/Jakarta EE.
* finagle: new, Twitter ecosystem
* armeria: new, Armeria ecosystem
* netty: new, mature underlying server, sparsely used http4s integration

## Clients

* blaze-client: oldest, most battle tested, no new features, some hard-to-reproduce bugs
* ember-client: only one without `ConcurrentEffect` in CE2, easiest to maintain, new features (e.g., H2) being added
* async-http-client: many people's recommendation. Melts under high load for me, others love it.
* okhttp-client: blocking, people speak ill of it lately, might have been better before Kotlin port?
* jettty-client: few users, few complaints.  I think it's neat.
* JavaNetClient: blocking, dubious resource management, no new dependencies, mostly intended for a REPL
* jdk-http-client: requires Java 11, no new dependencies, has a websocket client
* finagle, armeria, netty: similar to their server counterparts

## Summary

I agree with most of what @ScalaWilliam said, but Ember has the most development momentum and will become the "default" in the docs.  If you're just getting started, I'd start there.  But they all exist because someone had an itch to scratch, and it might be your itch, too.
